src/java.base/share/classes/java/io/Reader.java

```diff
@@ -184,12 +190,38 @@ protected Reader(Object lock) {
      * @since 1.5
      */
     public int read(CharBuffer target) throws IOException {
-        int len = target.remaining();
-        char[] cbuf = new char[len];
-        int n = read(cbuf, 0, len);
-        if (n > 0)
-            target.put(cbuf, 0, n);
-        return n;
+        int nread;
+        if (target.hasArray()) {
+            char[] cbuf = target.array();
+            int pos = target.position();
+            int rem = target.limit() - pos;
+            if (rem <= 0)
+                return -1;
+            int off = target.arrayOffset() + pos;
+            nread = this.read(cbuf, off, rem);
+            if (nread > 0)
+                target.position(pos + nread);
+        } else {
+            int remaining = target.remaining();
+            int bufferSize = Math.min(remaining, WORK_BUFFER_SIZE);
+            nread = 0;
+            synchronized (lock) {
+                if ((workBuffer == null) || (workBuffer.length < bufferSize))
+                    workBuffer = new char[bufferSize];
+                int n;
+                do {
+                    // read to EOF which may read more or less than buffer size
+                    if ((n = read(workBuffer, 0, Math.min(remaining, bufferSize))) > 0) {
+                        target.put(workBuffer, 0, n);
+                        nread += n;
+                        remaining -= n;
+                    }
+                    // if the last call to read returned -1 or the number of bytes
+                    // requested have been read then break
+                } while (n >= 0 && remaining > 0);
```

Comment:
The code for case that the char buffer has a backing array looks okay but I'm not sure about the direct buffer/other cases. One concern is that this is a read method, not a transferXXX method so we shouldn't be calling the underlying read several times. You'll see what I mean if you consider the scenario where you read < rem, then read again and the second read blocks or throws. I'l also concerned about "workBuffer" adding more per-stream footprint for cases where skip or read(CB) is used. Objects such as InputStreamReader are already a problem due to the underlying stream decoder.

Reply:
Right. So you propose to revert the off-heap path to the current master? That would be fine with me. The original bug and my motivation was only about the backing array case, the rest crept in. That would certainly keep the risk and impact lower.

Reply:
I think that's what @AlanBateman intended. The `skip()` changes would revert also but the C-style array changes can stay. Thanks.

Reply:
Yes, let's keep bring it back to just eliminating the intermediate array when the buffer has a backing array. The other case that been examined separated if needed but we can't use the approach proposed in the current PR because it changes the semantics of read when there is a short-read followed by a blocking or exception throwing read.

Reply:
Done